### PR TITLE
[Metric] Fix issue 10634

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

For shared metrics such as `RedisLatency`, it could have name conflicts. When names are conflicting, the metrics would have the "wrong  label keys", which causes the specified failure in the issue. For example, if the RedisLatency is from raylet, it should have 4 keys (component, version, node id, custom) whereas if it comes from gcs server, it should have 3 keys (component, version, custom).  

This PR resolves the issue by adding component to metric name. For example, if it comes from raylet, it will be `ray_raylet_redis_latency`.

## Related issue number

https://github.com/ray-project/ray/issues/10634

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by simon-mo** [comment]
Nice find! However, I don't think hacking the metric agent is the right way to do. Can we just add the node id for GCS server as a static label?

**Comment by rkooo567** [comment]
This can happen among any kind of shared metrics. The specified solution won't resolve the case where raylet and core worker share some metrics def (I assume this could happen easily. For example, imagine we are adding metrics on gcs server endpoints).

**Comment by rkooo567** [comment]
I think modifying name at metrics agent is fine because until metrics agent, it is implementation details.

**Comment by rkooo567** [comment]
Another solution is to modify the name when cpp accesses the metric for the first time (or when `stats::Init` is called). This solution might take some time to write code.

**Comment by simon-mo** [comment]
This is also weird for application metrics:`ray_core_worker_serve_router_latency`...

**Comment by rkooo567** [comment]
That's a good point. I think we should provide a way to specify namespace for application metrics (so that serve will be like ray_serve_[metric_name]).

**Comment by rkooo567** [comment]
But, I think this is reasonable solution for now (unless you have some other ideas. I think we can make custom metrics to have its own namespace later, and everything else is just implementation details, so we can change when we have more bandwidth) especially given the product team needs a fix for this issue.

**Comment by simon-mo** [comment]
actually, shouldn't it be `gcs_latency` in raylet and core worker. `redis` is just an implementation detail for GCS. I think it is the responsibility for the metric users to make sure shared metrics have the same key. After all, there is a reason for them to have the same name.

**Comment by rkooo567** [comment]
Yeah, we will change it to `gcs_latency` (we have a doc to aggregate plans to change metrics def). 

Btw, now I understand what you meant. So, are you suggesting to add node id label to GCS, so that it can have the same length lables?

**Comment by rkooo567** [comment]
I feel like this solution is so much more cumbersome though. With the current solution, there's no need to have matching label keys. What's the disadvantage of the current approach, and the benefit of your approach?

**Comment by rkooo567** [comment]
The naming pattern `ray_[component]_[metric_name]` is also something we'd like to encourage people to use. (https://github.com/ray-project/ray/blob/840fb5543b6ffca119cfb080b6cfe62060893f54/python/ray/experimental/metrics.py#L29) (btw, note that prom doesn't support `.` delimeter, so we will change this comment to suggest `_` delimeter)

**Comment by simon-mo** [comment]
I was trying to propose not to check the component in exporter. Instead, we just rename the metrics to:
- `gcs_latency{component, version, node id, custom}` from Raylet and Core worker
- `redis_latency{component, version, custom}` from GCS.

This is because `gcs_latency` measures the latency to talk the gcs server, and `redis_latency` measures the time for gcs_server <-> redis server. These two metrics are simply different.

The benefits are:
1. clear semantics
2. no need to special case "Component" key: making cleaner metric_agent code.
3. no need to force prefix component as part of metric name: making cleaner more meaningful metric names.

One possible downside is that future uses of colliding metric names with different label keys can silently error. There are two points about this:
1. We should catch the error and print an error message to driver in metric agent if possible. Prometheus endpoint should not crash.
2. It is absolutely user's responsibility to ensure tag keys are the same when using the same metric name. This is because given a name, metric measures *a thing*. Labels are metadata for *that thing* to be used to drill down and roll up metrics. https://prometheus.io/docs/practices/naming/

**Comment by rkooo567** [comment]
> This is because given a name, metric measures a thing. Labels are metadata for that thing to be used to drill down and roll up metrics

This makes a lot of sense. I will follow your suggestion but will have one more modification. I will have the same key length for every component (gcs_server, core_worker, raylet) because there will be shared metrics among all components (grpc_server endpoint load). Thanks for the good writeup :)

**Comment by rkooo567** [comment]
Thinking more, I feel like the original approach was the right choice. I have impression if the same name metric is shared by all other components, they should have different name, than we distinguish them by tags. It is a little bit weird that we should have the same global tags for all components. What about I write a doc, and we discuss this in the meeting?

**Comment by simon-mo** [comment]
Sure

**Comment by simon-mo** [comment]
@rkooo567 can you share the doc and post the result here?

**Comment by rkooo567** [comment]
Doc: https://docs.google.com/document/d/1g5WRpbdzbnGnR0_9-6llNHd0jo_Eo2KZ998R2b9-0wA/edit#

Conclusion: 
- Going with the Simon's proposal. If something turns out to be bad, switching to the other one is trivial.
- Have the same global keys for all components (component, version, ip_address).

Will work on it later this week (or whenever I have time outside work hours. This is not on my task).

**Comment by rkooo567** [comment]
@simon-mo Can we merge this first to unblock product issues?

File: src/ray/gcs/gcs_server/gcs_server_main.cc [hunk 1]
```diff
@@ -59,8 +59,9 @@ int main(int argc, char *argv[]) {
   }
 
   RayConfig::instance().initialize(config_map);
-  const ray::stats::TagsType global_tags = {{ray::stats::ComponentKey, "gcs_server"},
-                                            {ray::stats::VersionKey, "1.1.0.dev0"}};
+  const ray::stats::TagsType global_tags = {
+    {ray::stats::ComponentKey, "gcs_server"}, {ray::stats::VersionKey, "1.1.0.dev0"},
+    {ray::stats::NodeAddressKey, node_ip_address};
```
Comment (simon-mo) [inline src/ray/gcs/gcs_server/gcs_server_main.cc:64]: Miss the closing }

File: src/ray/gcs/gcs_server/gcs_server_main.cc [hunk 2]
```diff
@@ -59,8 +59,9 @@ int main(int argc, char *argv[]) {
   }
 
   RayConfig::instance().initialize(config_map);
-  const ray::stats::TagsType global_tags = {{ray::stats::ComponentKey, "gcs_server"},
-                                            {ray::stats::VersionKey, "1.1.0.dev0"}};
+  const ray::stats::TagsType global_tags = {
+    {ray::stats::ComponentKey, "gcs_server"}, {ray::stats::VersionKey, "1.1.0.dev0"},
```
Comment (simon-mo) [inline src/ray/gcs/gcs_server/gcs_server_main.cc:63]: Lint

**Review by simon-mo — APPROVED** [review]

File: src/ray/gcs/gcs_server/gcs_server_main.cc (hunk 1 above)
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_server_main.cc:64]: oops.. thanks for cathing it ;)

**Pull request merged**